pkg/apis/org/conversion.go

```diff
@@ -0,0 +1,193 @@
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	v1StorageAnnotation          = "che.eclipse.org/cheClusterV1Spec"
+	v2alpha1StorageAnnotation    = "che.eclipse.org/cheClusterV2alpha1Spec"
+	routeDomainSuffixPropertyKey = "CHE_INFRA_OPENSHIFT_ROUTE_HOST_DOMAIN__SUFFIX"
+)
+
+func V1ToV2alpha1(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) error {
+	v2Data := v1.Annotations[v2alpha1StorageAnnotation]
+	v2Spec := v2alpha1.CheClusterSpec{}
+	if v2Data != "" {
+		err := yaml.Unmarshal([]byte(v2Data), &v2Spec)
+		if err != nil {
+			return err
+		}
+	}
+
+	v2.ObjectMeta = v1.ObjectMeta
+	v2.Spec = v2Spec
+
+	v1Spec, err := yaml.Marshal(v1.Spec)
+	if err != nil {
+		return err
+	}
+	if v2.Annotations == nil {
+		v2.Annotations = map[string]string{}
+	}
+	v2.Annotations[v1StorageAnnotation] = string(v1Spec)
+
+	v1ToV2alpha1_Host(v1, v2)
+	v1ToV2alpha1_WorkspaceBaseDomain(v1, v2)
+	v1ToV2alpha1_GatewayEnabled(v1, v2)
+	v1ToV2alpha1_GatewayImage(v1, v2)
+	v1ToV2alpha1_GatewayConfigurerImage(v1, v2)
+	v1ToV2alpha1_TlsSecretName(v1, v2)
+	v1ToV2alpha1_K8sIngressAnnotations(v1, v2)
+
+	return nil
+}
+
+func V2alpha1ToV1(v2 *v2alpha1.CheCluster, v1Obj *v1.CheCluster) error {
+	v1Data := v2.Annotations[v1StorageAnnotation]
+	v1Spec := v1.CheClusterSpec{}
+	if v1Data != "" {
+		err := yaml.Unmarshal([]byte(v1Data), &v1Spec)
+		if err != nil {
+			return err
+		}
+	}
+
+	v1Obj.ObjectMeta = v2.ObjectMeta
+	v1Obj.Spec = v1Spec
+	v1Obj.Status = v1.CheClusterStatus{}
+
+	v2Spec, err := yaml.Marshal(v2.Spec)
+	if err != nil {
+		return err
+	}
+	if v1Obj.Annotations == nil {
+		v1Obj.Annotations = map[string]string{}
+	}
+	v1Obj.Annotations[v2alpha1StorageAnnotation] = string(v2Spec)
+
+	v2alpha1ToV1_Host(v1Obj, v2)
+	v2alpha1ToV1_WorkspaceBaseDomain(v1Obj, v2)
+	v2alpha1ToV1_GatewayEnabled(v1Obj, v2)
+	v2alpha1ToV1_GatewayImage(v1Obj, v2)
+	v2alpha1ToV1_GatewayConfigurerImage(v1Obj, v2)
+	v2alpha1ToV1_TlsSecretName(v1Obj, v2)
+	v2alpha1ToV1_K8sIngressAnnotations(v1Obj, v2)
+
+	return nil
+}
+
+func v1ToV2alpha1_Host(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) {
+	v2.Spec.Gateway.Host = v1.Spec.Server.CheHost
+}
+
+func v1ToV2alpha1_WorkspaceBaseDomain(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) {
+	if util.IsOpenShift {
+		v2.Spec.WorkspaceBaseDomain = v1.Spec.Server.CustomCheProperties[routeDomainSuffixPropertyKey]
+	} else {
+		v2.Spec.WorkspaceBaseDomain = v1.Spec.K8s.IngressDomain
+	}
+}
+
+func v1ToV2alpha1_GatewayEnabled(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) {
+	val := util.GetServerExposureStrategy(v1) == "single-host"
+	v2.Spec.Gateway.Enabled = &val
+}
+
+func v1ToV2alpha1_GatewayImage(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) {
+	v2.Spec.Gateway.Image = v1.Spec.Server.SingleHostGatewayImage
+}
+
+func v1ToV2alpha1_GatewayConfigurerImage(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) {
+	v2.Spec.Gateway.ConfigurerImage = v1.Spec.Server.SingleHostGatewayConfigSidecarImage
+}
+
+func v1ToV2alpha1_TlsSecretName(v1 *v1.CheCluster, v2 *v2alpha1.CheCluster) {
```

Comment:
How it works know, for instance: 
* OpenShift domain is `my.openshift.com`
* User is user is able to specify `spec.server.cheHost: my-ide.com` (something different from OpenShift domain)
* `spec.server.cheHostTLSSecret` is used to secure that domain.

So, how it will work after changes? Will user able to set and secure any Che host we wants?

Reply:
To be fair, the situation is not that bad in DWCO. DWCO only deals with workspace endpoints, so for example the dashboard URL is not its concern. But still we need to support `domain/subpath` endpoints as well as `subdomain.domain`. To be able to use a different certificate for each, I added the capability to the model.